libnd4j/include/ops/declarable/generic/random/multinomial.cpp

```diff
@@ -0,0 +1,115 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+// @author Oleh Semeniv (oleg.semeniv@gmail.com)
+//
+
+#include <op_boilerplate.h>
+#if NOT_EXCLUDED(OP_random_multinomial)
+
+#include <ops/declarable/CustomOperations.h>
+#include <helpers/RandomLauncher.h>
+#include <ops/declarable/helpers/random.h>
+
+namespace nd4j {
+    namespace ops {
+        ///////////////////////
+        /**
+         * multinomial (categorical) random generator
+         * takes 2D ndarray with logits with shape [batch_size (N), num_classes (K)]
+         * represents the unnormalized log-probabilities for all classes.
+         * Int arguments: 0 - scalar value of samples number, number of independent samples to draw for each experiment 1,N.
+         * Int arguments: 1 - optional argument, corresponds to dimension with batch_size
+         * Int arguments: 2 - optional argument, integer type to use for the output. Default int64.
+         * Int arguments: 3 - optional argument, integer used to create a random seed for the distribution
+         */
+         // used https://en.wikipedia.org/wiki/Categorical_distribution
+         // methods: gumbel trick + softmax + argmax
+        CUSTOM_OP_IMPL(random_multinomial, 1, 1, false, 0, 1) {
+            
+            auto input = INPUT_VARIABLE(0);
+            auto output = OUTPUT_VARIABLE(0);
+            auto nSamples = INT_ARG(0);
+            const int argSize = block.getIArguments()->size();
+
+            REQUIRE_TRUE(argSize >= 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
+                                          " but got no argumets instead.");
+
+            // do nothing if number of samples = 0
+            if (0 == nSamples) 
+                return Status::OK();
+
+            REQUIRE_TRUE(!input->isEmpty(), 0, "RANDOM_MULTINOMIAL OP: Have to be provided at least one logits. ");
+            REQUIRE_TRUE(nSamples > 0, 0, "RANDOM_MULTINOMIAL OP: Number of samples should be greater then 0, got %i. ", nSamples);
+
+            const int rank = input->rankOf();
+            REQUIRE_TRUE(rank == 2, 0, "RANDOM_MULTINOMIAL OP: Logits should be a matrix with rank = 2, but got instead rank = %i.", rank);
+            const int dimC = argSize > 1 ? (INT_ARG(1) >= 0 ? INT_ARG(1) : INT_ARG(1) + rank) : rank - 1;
+
+            auto dimA = (0 == dimC) ? 1 : 0;
+            if (1 == input->sizeAt(dimA)) {
+                *output = 0;
+                return Status::OK();
+            }
+
+            auto rng = block.randomGenerator();
+            if (argSize > 3) {
+                rng.setStates(static_cast<Nd4jLong>(INT_ARG(3)), static_cast<Nd4jLong>(INT_ARG(3)));
```

Comment:
As discussed before - don't do this please. Just use RNG as is, and assume it's propagated properly.

Reply:
P.s. for tests purposes use Context directly, and set both rng states there.

Reply:
done